### ISSUE TITLE
QUESTION: How can I specify ACF fields in an ep_weighting_fields_for_post_type filter?

### ISSUE THREAD

**Issue by aronvandervalk** [10up/ElasticPress#2798]
Hi,

I am trying to implement custom weighting, using ACF fields. 
We are using the WP_Query to query a specific set of posts:

```
$args = array(
  'post_type' => [$context['post_type']],
  's' => $context['search_input'],
  'meta_query' => $meta_query,
  'tax_query' => $tax_query,
  'search_fields' => array(
    'meta' => array(
      'introduction_title',
      'introduction_training_code',
      'short_description',
      'full_description'
    )
  ),
);
```

[This article](https://www.elasticpress.io/blog/2020/06/custom-fields-and-weighted-search-with-elasticpress-part-2/) explains how to the add the custom field to the weighting settings. 
```
// Change my_custom_field here to what you need.
$key = 'meta.my_custom_field.value';

$fields['meta']['children'][ $key ] = array(
	'key'   => $key,
	'label' => __( 'My Custom Field', 'textdomain' ),
);
```

We can successfully add these fields to the settings page, but making them searchable and giving them a weight does not seem to alter the results. 
![CleanShot 2022-05-25 at 12 57 33](https://user-images.githubusercontent.com/326355/170247047-975961b8-df74-43d8-be5d-ca430de51b12.png)

What should be the exact value of `$key = 'meta.my_custom_field.value';` if the fields name is in this case `introduction_title`? 
I have tried  `meta.introduction_title.value`,  `meta.introduction_title` and `introduction_title`, but none seem to alter the results. Also after re-indexing

Looking at the query from the debug bar, I am guessing it should be `meta.introduction_title.value`, but this does not seem to alter the results when querying the WP_Query.

**Comment by felipeelia** [comment]
Hi @aronvandervalk,

It should be `meta.introduction_title.value`. It seems you are already using Debug Bar ElasticPress, right? If so, do you mind sharing an example of an ES query you are sending and an example of an Elasticsearch document that should match? For the latter, you can either search for something that brings results and share the "Query Results" content or run `wp elasticpress request /<post-index-name>/_doc/<pos_id> --pretty`. You can get <post-index-name> in the Index Health submenu page.

**Comment by aronvandervalk** [comment]
Hi @felipeelia

Thanks for the reply.

When checking the debug bar I see two ElasticPress queries:
The first request body:
```
{
    "from": 0,
    "size": 10,
    "sort": [
        {
            "post_date": {
                "order": "desc"
            }
        }
    ],
    "query": {
        "function_score": {
            "query": {
                "bool": {
                    "should": [
                        {
                            "multi_match": {
                                "query": "Excel",
                                "type": "phrase",
                                "fields": [
                                    "meta.introduction_title.value",
                                    "meta.introduction_training_code.value",
                                    "meta.introduction_certification_code.value",
                                    "meta.short_description.value",
                                    "meta.full_description.value"
                                ],
                                "boost": 3
                            }
                        },
                        {
                            "multi_match": {
                                "query": "Excel",
                                "fields": [
                                    "meta.introduction_title.value",
                                    "meta.introduction_training_code.value",
                                    "meta.introduction_certification_code.value",
                                    "meta.short_description.value",
                                    "meta.full_description.value"
                                ],
                                "boost": 1,
                                "fuzziness": "auto",
                                "operator": "and"
                            }
                        },
                        {
                            "multi_match": {
                                "query": "Excel",
                                "type": "cross_fields",
                                "fields": [
                                    "meta.introduction_title.value",
                                    "meta.introduction_training_code.value",
                                    "meta.introduction_certification_code.value",
                                    "meta.short_description.value",
                                    "meta.full_description.value"
                                ],
                                "boost": 1,
                                "analyzer": "standard",
                                "tie_breaker": 0.5,
                                "operator": "and"
                            }
                        }
                    ]
                }
            },
            "functions": [
                {
                    "exp": {
                        "post_date_gmt": {
                            "scale": "14d",
                            "decay": 0.25,
                            "offset": "7d"
                        }
                    }
                },
                {
                    "weight": 0.001
                }
            ],
            "score_mode": "sum",
            "boost_mode": "multiply"
        }
    },
    "post_filter": {
        "bool": {
            "must": [
                {
                    "bool": {
                        "must": [
                            {
                                "terms": {
                                    "meta.introduction_active.raw": [
                                        "1"
                                    ]
                                }
                            },
                            {
                                "terms": {
                                    "meta.introduction_searchable.raw": [
                                        "1"
                                    ]
                                }
                            }
                        ]
                    }
                },
                {
                    "terms": {
                        "post_type.raw": [
                            "course"
                        ]
                    }
                },
                {
                    "term": {
                        "post_status": "publish"
                    }
                }
            ]
        }
    }
}
```
ThIs request returns the items in the list, but these items are not affected by the weighting.

=======

The second query does seem to take the weight into account, but this query does not return any results.
```
"meta.introduction_title.value^22",
"meta.introduction_training_code.value^100",
```

Body:
```
{
    "from": 0,
    "size": 10,
    "sort": [
        {
            "_score": {
                "order": "desc"
            }
        }
    ],
    "query": {
        "function_score": {
            "query": {
                "bool": {
                    "should": [
                        {
                            "bool": {
                                "must": [
                                    {
                                        "bool": {
                                            "should": [
                                                {
                                                    "multi_match": {
                                                        "query": "ep_autosuggest_placeholder",
                                                        "type": "phrase",
                                                        "fields": [
                                                            "meta.full_description.value^1",
                                                            "meta.introduction_title.value^22",
                                                            "meta.introduction_training_code.value^100",
                                                            "meta.short_description.value^1",
                                                            "terms.ep_custom_result.name^9999"
                                                        ],
                                                        "boost": 3
                                                    }
                                                },
                                                {
                                                    "multi_match": {
                                                        "query": "ep_autosuggest_placeholder",
                                                        "fields": [
                                                            "meta.full_description.value^1",
                                                            "meta.introduction_title.value^22",
                                                            "meta.introduction_training_code.value^100",
                                                            "meta.short_description.value^1"
                                                        ],
                                                        "boost": 1,
                                                        "fuzziness": "auto",
                                                        "operator": "and"
                                                    }
                                                },
                                                {
                                                    "multi_match": {
                                                        "query": "ep_autosuggest_placeholder",
                                                        "type": "cross_fields",
                                                        "fields": [
                                                            "meta.full_description.value^1",
                                                            "meta.introduction_title.value^1",
                                                            "meta.introduction_training_code.value^1",
                                                            "meta.short_description.value^1",
                                                            "terms.ep_custom_result.name^1"
                                                        ],
                                                        "boost": 1,
                                                        "analyzer": "standard",
                                                        "tie_breaker": 0.5,
                                                        "operator": "and"
                                                    }
                                                }
                                            ]
                                        }
                                    }
                                ],
                                "filter": [
                                    {
                                        "match": {
                                            "post_type.raw": "course"
                                        }
                                    }
                                ]
                            }
                        },
                        {
                            "bool": {
                                "must": [
                                    {
                                        "bool": {
                                            "should": [
                                                {
                                                    "multi_match": {
                                                        "query": "ep_autosuggest_placeholder",
                                                        "type": "phrase",
                                                        "fields": [
                                                            "post_title^1",
                                                            "terms.certification_category.name^1",
                                                            "terms.ep_custom_result.name^9999"
                                                        ],
                                                        "boost": 3
                                                    }
                                                },
                                                {
                                                    "multi_match": {
                                                        "query": "ep_autosuggest_placeholder",
                                                        "fields": [
                                                            "post_title^1",
                                                            "terms.certification_category.name^1",
                                                            "post_title.suggest^1",
                                                            "term_suggest^1"
                                                        ],
                                                        "boost": 1,
                                                        "fuzziness": "auto",
                                                        "operator": "and"
                                                    }
                                                },
                                                {
                                                    "multi_match": {
                                                        "query": "ep_autosuggest_placeholder",
                                                        "type": "cross_fields",
                                                        "fields": [
                                                            "post_title^1",
                                                            "terms.certification_category.name^1",
                                                            "terms.ep_custom_result.name^1"
                                                        ],
                                                        "boost": 1,
                                                        "analyzer": "standard",
                                                        "tie_breaker": 0.5,
                                                        "operator": "and"
                                                    }
                                                }
                                            ]
                                        }
                                    }
                                ],
                                "filter": [
                                    {
                                        "match": {
                                            "post_type.raw": "certification"
                                        }
                                    }
                                ]
                            }
                        }
                    ]
                }
            },
            "functions": [
                {
                    "exp": {
                        "post_date_gmt": {
                            "scale": "14d",
                            "decay": 0.25,
                            "offset": "7d"
                        }
                    }
                },
                {
                    "weight": 0.001
                }
            ],
            "score_mode": "sum",
            "boost_mode": "multiply"
        }
    },
    "post_filter": {
        "bool": {
            "must": [
                {
                    "terms": {
                        "post_type.raw": [
                            "post",
                            "page",
                            "course",
                            "certification",
                            "job",
                            "person",
                            "location"
                        ]
                    }
                },
                {
                    "terms": {
                        "post_status": [
                            "publish",
                            "acf-disabled"
                        ]
                    }
                }
            ]
        }
    }
}
```

Also, I do not understand where this query is coming from. When temporarily disabling the WP_Query request and the search input or even all page elements, the second request still occurs. Could that be due to the fact this particular page is a taxonomy template page?

**Comment by felipeelia** [comment]
Hi @aronvandervalk,

The second query is created by the Autosuggest feature as a model to be used by the AJAX functionality. Although it appears in all page loads, ElasticPress caches its results, so it is not actually fired every time.

For the first one, you are not using `search_fields`, right? It might help if you can share all custom code you have related to ElasticPress. Also, are you using it in a multisite? If so, is EP network activated? Thanks!

**Comment by aronvandervalk** [comment]
Hi @felipeelia 

I see that the second query is indeed autosuggest, which can be removed by disabling the autosuggest setting.

Leaves us with the first query. Our setup is that we want to combine meta queries, tax queries _and_ a custom search on `search_fields` in a Query setup. This works, but the weighting is not taken into account (which does seem to appear in the autosuggest query). 
- No multi site
- Not much custom code, only one `ep_prepare_meta_data` filter is used to filter out fields to skip indexing

When opening the ElasticPress debugbar we see the body as posted in the first request body above.

**Comment by felipeelia** [comment]
Hi @aronvandervalk,

As you are using ` ep_weighting_fields_for_post_type` -- the filter pointed out in that article -- you should NOT use `search_fields` at all, as that will overwrite the weighting settings completely. Remove the piece of code where you are adding the `search_fields` clause to WP_Query and ElasticPress should inject into it all the fields marked as searchable in the Weighing Dashboard.

If you have a scenario where you want to have different sets of weighted fields, then you'll want to use the solution pointed out in the first part of that blog post: https://www.elasticpress.io/blog/2020/04/custom-fields-and-weighted-search-with-elasticpress/

In summary:
- `search_fields` in WP_Query: no weighting will be used
- `ep_weighting_configuration_for_search`: hard-coded weighting
- `ep_weighting_fields_for_post_type`: integrate with the Weighting Dashboard

Can you please give it a try and let us know if this fixes the problem you are facing? Thanks!

**Comment by aronvandervalk** [comment]
Hi @felipeelia 

Thank you very much for the reply. Now it makes sense. For me it was unclear in the [example](https://www.elasticpress.io/blog/2020/06/custom-fields-and-weighted-search-with-elasticpress-part-2/)
that when search_fields in WP_Query are set, no weighting will be applied.

When removing them from the WP_Query the setup works as expected.

Thanks!!
